src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,331 @@
+    po.Register("snr-db", &snr_db,
+                "desired SNR(dB) of the output");
+
+    po.Read(argc, argv);
+    if (po.NumArgs() != 3) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    std::string input_wave_file = po.GetArg(1);
+    std::string rir_file = po.GetArg(2);
+    std::string output_wave_file = po.GetArg(3);
+
+    WaveData input_wave;
+    {
+      Input ki(input_wave_file);
+      input_wave.Read(ki.Stream());
+    }
+
+    const Matrix<BaseFloat> &input_matrix = input_wave.Data();
+    BaseFloat samp_freq_input = input_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_input = input_matrix.NumCols(),  // number of samples in recording
+          num_chan_input = input_matrix.NumRows();  // number of channels in recording
+
+    KALDI_LOG << "samp_freq of source is " << samp_freq_input
+              << " num_samp " << num_samp_input
+              << " num_chan " << num_chan_input;
+
+    KALDI_ASSERT(input_wave_channel < num_chan_input);
+
+    WaveData rir_wave;
+    {
+      Input ki(rir_file);
+      rir_wave.Read(ki.Stream());
+    }
+    const Matrix<BaseFloat> &rir_matrix = rir_wave.Data();
+    BaseFloat samp_freq_rir = rir_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_rir = rir_matrix.NumCols(),  // number of samples in recording
+          num_chan_rir = rir_matrix.NumRows();  // number of channels in recording
+    KALDI_LOG << "samp_freq of rir is " << samp_freq_rir
+               << " num_samp " << num_samp_rir
+               << " num_chan " << num_chan_rir;
+    if (!multi_channel_output) {
+      KALDI_ASSERT(rir_channel < num_chan_rir);
+    }
+
+    Matrix<BaseFloat> noise_matrix;
+    if (!noise_file.empty()) {
+      WaveData noise_wave;
+      {
+        Input ki(noise_file);
+        noise_wave.Read(ki.Stream());
+      }
+      noise_matrix = noise_wave.Data();
+      BaseFloat samp_freq_noise = noise_wave.SampFreq();
+      int32 num_samp_noise = noise_matrix.NumCols(),
+            num_chan_noise = noise_matrix.NumRows();
+      KALDI_LOG << "samp_freq of noise is " << samp_freq_noise
+                << " num_samp " << num_samp_noise
+                << " num_chan " << num_chan_noise;
+      if (multi_channel_output) {
+        KALDI_ASSERT(num_chan_rir == num_chan_noise);
+      } else {
+        KALDI_ASSERT(noise_channel < num_chan_noise);
+      }
+    }
+
+    int out_chan = 0;
+    int rir_chan = rir_channel;
+    int noise_chan = noise_channel;
+    int num_chan_output = 1;
+    if (multi_channel_output) {
+      num_chan_output = num_chan_rir;
+      rir_chan = 0;
+      noise_chan = 0;
+    }
+
+    Matrix<BaseFloat> out_matrix(num_chan_output, num_samp_input);
+
+    do {
+      Vector<BaseFloat> input(num_samp_input);
+      input.CopyRowFromMat(input_matrix, input_wave_channel);
+      Vector<BaseFloat> rir(num_samp_rir);
+      rir.CopyRowFromMat(rir_matrix, rir_chan);
+      rir.Scale(1 / static_cast<float>(2147483647));  // Assume 32 bits 
```

Comment:
Vijay, does the absolute scaling of the rir even matter?  Is it supposed to be normalized somehow?  I am a bit confused.